# PROYECTO FINAL

## LIBRERIAS

### IMPORTACION DE LIBRERIAS

In [156]:
import pandas as pd
import numpy as np
import os

## FUNCIONES

### CAMBIO CAMPO A TIPO DATATIME

In [158]:
def CambioTipoFecha(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    df[columnas] = df[columnas].apply(pd.to_datetime) # aplicamos a las columnas obtenidas el cambio de tipo de dato a datatime
    return df # Retornamos el dataframe con el cambio de tipo de dato en las clumnas 

### CAMBIO CAMPO A MINUSCULA

In [160]:
def CambioMinuscula(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    for columna in columnas: # trabajamos con un for para seleccionar columna por columna deseadas, ya que str.lower trabaja por serie y no por dataframe
        df[columna] = df[columna].str.lower() # aplicamos str.lower, el cual hace que todos los valores aparezcan en minuscula
    return df # retornamos el dataframe modificado

### CAMBIO CAMPO A MAYUSCULA

In [162]:
def CambioMayuscula(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    for columna in columnas: # trabajamos con un for para seleccionar columna por columna deseadas, ya que str.lower trabaja por serie y no por dataframe
        df[columna] = df[columna].str.upper() # aplicamos str.upper, el cual hace que todos los valores aparezcan en mayuscula 
    return df # retornamos el dataframe modificado

### CAMBIO CAMPO A TIPO STR

In [164]:
def CambioTipoStr(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    df[columnas] = df[columnas].astype("string") # aplicamos a las columnas obtenidas el cambio de tipo de dato a str
    return df # Retornamos el dataframe con el cambio de tipo de dato en las clumnas

## EXTRACCIÓN

### CARGA

In [273]:
# En todas estas lineas, lo que estamos haciendo es leer harchivos con formato CSV, los cuales se cuentran en una carpeta de un repositorio
# de github, y guardarlos cono dataframe mediante la libre ria pandas 

df_customers = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")
df_orders = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")
df_order_items = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")
df_products = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")

### VISUALIZACIÓN

REALIZAREMOS UNA VISUALIZACION DE LOS 5 DATAFRAME, CON EL FIN DE PODER OBSERVAR

QUE TIPO DE DATOS ENCONTRAMOS EN CADA UNO DE ELL0S

#### DATAFRAME DF_CUSTOMERS

In [170]:
df_customers.head() # Visalizamos los primeros 5 registros del dataframe df_customers, con el fin de poder ver que campos contiene y tipos de datos

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [174]:
df_customers.shape # Con shape lo que hacemos, es visualizar la estructura del dataframe, con el fin de ver cuantos registros y campos hay

(99441, 5)

In [176]:
df_customers.dtypes # Con dtypes, visualizamos el tipo de dato que encontramos en cada campo

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [178]:
df_customers.isnull().sum() # con isnull.sum, realizamos un conteo de cantidad de valores nulos en cada campo

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [180]:
df_customers.duplicated().sum() # Con duplicated.sum, realizamos un conteo de valores duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_CUSTOMERS, QUE PRESETA 5 CAMPOS Y 99441 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO Y DE TIPO ENTERO, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODES INFORMAR QUE NO PRESENTA DATOS NULOS, POR LO CUAL NO SERA NECESARIO TRATAR DICHO TEMA Y TAMPOCO CONTIENE REGISTROS DUPLICADOS

#### DATAFRAME DF_ORDERS

In [183]:
df_orders.head() # visualizamos los primero 5 registros del dataframe df_orders, con el fin de ver que campos y tipo de datos hay

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [185]:
df_orders.shape # con shape vemos la estructura del dataframe, pasa saber cuantos registros y campos hay

(99441, 8)

In [187]:
df_orders.dtypes # con dtype vemos que tipos de datos tenemos en los campos del dataframe

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [189]:
df_orders.isnull().sum() # con isnull.sum podemos ver la cantidad de valores nulos por cada campo

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [191]:
df_orders.duplicated().sum() # Con duplicated.sum podemos visualizar cuantos calores duplicados hay

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDERS, QUE PRESETA 8 CAMPOS Y 99441 REGISTROS

TENEMOS TODOS LOS CAMPOS DE TIPO OBJETO, PERO SE OBSERVAN REGISTOS DE FECHAS, POR LO QUE SERA NECESARIO

PASARLOS A TIPO DATE. TAMBIEN SERA ESTUDIADO SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE PARA LOS CAMPOS RESTANTES

POR ULTIMO PODEMOS INFORMAR QUE PRESENTA REGISTROS NULOS, EN TRES CAMPOS, LOS CUAL DEBERÁ SER TRATADO MAS ADELANTE, EVALUANDO CUAL ES LA ACCIÓN

MAS CONVENIENTE PARA ESTE CASO Y TAMBIÉN PODEMOS DECIR QUE NO PRESENTA DUPLICADOS

#### DATAFRAME DF_ORDER_ITEMS

In [194]:
df_order_items.head() # Visualizamos los primeros 5 registros del dataframe df_order_items, con el fin de poder ver que camposy datos tenemos

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [196]:
df_order_items.shape # Con shape estamos viendo la estructura del dataframe para saber cuando registros y campos hay

(112650, 7)

In [198]:
df_order_items.dtypes # Con dtype estamos viendo que tipos de datos tenemos en el dataframe

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [200]:
df_order_items.isnull().sum() # en esta linea estamos contando la cantidad de valores nuloes q tenemos en cada campo

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [202]:
df_order_items.duplicated().sum() # En esta lonea estamos visualizando cuando duplicados existen

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDER_ITEMS, QUE PRESETA 7 CAMPOS Y 112650 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO, DE TIPO ENTERO Y DE TIPO FLOAT, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE. YA SE PUEDE DECIR QUE EL CAMPO QUE CONTIENE FECHAS, QUE ES DE TIPO OBJETO

TENDRA QUE SER SETEADO A TIPO DATE

POR ULTIMO PODEMOS INFORMAR QUE NO PRESENTA REGISTROS NULOS Y TAMPOCO TIENE DUPLICADOS

#### DATAFRAME DF_PRODUCTS

In [205]:
df_products.head() # Con head visualizamos los primeros 5 registros para poder ver que campos y q variables nos encontrmos

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [207]:
df_products.shape # en esta linea buscamos visualizar cual es la estructura del dataframe, para saber cuantos campos y regustros existen

(32951, 9)

In [209]:
df_products.dtypes # con dtypes buscamos saber que tipo de variables existen en cada campo

product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [211]:
df_products.isnull().sum() # visualizamos la cantidad de valores nulos en cada campo

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [213]:
df_products.duplicated().sum() # de esta manera contamos la cantidad de duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_PRODUCTS, QUE PRESENTA 9 CAMPOS Y 32951 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO Y DE TIPO FLOAT, LOS CUALES MAS ADELANTE SERAN ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODEMOS VER QUE EL DATAFRAME PRESENTA REGISTROS NULOS EN TODOS SUS CAMPOS, MENOS EN UNO, LO CUAL SERA TRATADO MAS ADELANTE,

APLICANDO EL METODO MAS CONVENIENTE PARA ESTE CASO Y TAMBIEN PODEMOS INFORMA QUE NO PRESENTA DUPLICADOS

#### DATAFRAME DF_ORDER_PAYMENTS

In [216]:
df_order_payments.head() # En esta linea buscamos visualizar los primeros 5 registros, con el fin de ver los campos y q variables existen

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [218]:
df_order_payments.shape # En esta linea se consulta cual es la estructura del dataframe, con el fin de ver la cantidad de campos y registros

(103886, 5)

In [220]:
df_order_payments.dtypes # con dtypes consultamos el tipo de variable que existe en cada campo

order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

In [222]:
df_order_payments.isnull().sum() # aca estamos contando la cantidad de valores nulos que existen en cada campo

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [224]:
df_order_payments.duplicated().sum() # con esta linea contamos la cantidad de duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDER_PAYMENTS, QUE PRENSETA 5 CAMPOS Y 103886 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO, DE TIPO ENTERO Y TIPO FLOAT, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODEMOS INFORMAR QUE NO SE PRESENTAN REGISTROS NULOS Y TAMPOCO DUPLICADOS.

## TRANSFORMACIÓN

###  DF_ORDERS

#### ANALIZAMOS VALORES NULOS

PARA EL TRATAMIENTO DE LOS NULOS EN ESTE DATAFRAME, TENDREMOS EN CONCIDERACION QUE MANTENDREMOS AQUELAS COLUMNAS QUE NO SUPEREN

UN 5% DE VALORES NULOS, EN EL TOTAL DE VALORES. UNA VEZ QUE TENGAMOS EL DATAFRAME NUEVO CON LAS COLUMNAS QUE COMPLAN CON DICHA

CONDICION, SE PROCEDERÁ A REALIZAR LA ELIMINACION DE LOS REGISTROS QUE CONTENGAN VALORES NULOS.

In [2]:
# En la linea siguiente lo que estamos haciendo es obtener de manera porcentual, respecto a la antidad total de registros, cual es el
# porcentaje de valore nulos en cada campo.
porcentaje_nulos = (df_orders.isnull().sum())*100/(len(df_orders))
porcentaje_nulos

# Podemos ver cuales son los porcentajes de nulosque existe en cada campo. En la linea siguiente vamos a filtrar por porcentaje, es decir
# los camposque tengan un porcentaje de nulos menor a 5% seran campos considerados, si supera al 5%, esos campos no seran tenidos en cuenta.
# Los campos se encuentran en el index, es por eso que estamos guardando en columnas, las columnas que complen con la condicion explicada.
# Elcampo que complua con la condicion del 5% y contenga valores nulos, esos registros seran eliminados
columnas = porcentaje_nulos[porcentaje_nulos  < 5].index
df_orders = df_orders[columnas] # en esta linea estamos modificando el dataframe para que contenga las clumnas deseadas
df_orders = df_orders.dropna() # en esta linea estamos eliminado los registros que contienen valores nulos.

NameError: name 'df_orders' is not defined

#### CAMBIAMOS CAMPOS A TIPO DATATIME

In [275]:
# En esta celda vamos a convertir el tipo de datos de algunos campos a tipo datatime, para eso en la siguiente linea lo que hcemos es
# crear una lista con los campos deseados.
columnas = ["order_purchase_timestamp","order_approved_at","order_delivered_carrier_date","order_delivered_customer_date","order_estimated_delivery_date"]
# en la siguiente linea llamamos a la funcon que se encarga de de realizar el cambio de tipo de variable, en la que le enviamos el dataframe que se quiere
# modificar y tambien la lista con las columnas que deseamos modificar
df_orders_mod = CambioTipoFecha(df_orders,columnas)

#### CAMBIAMOS CAMPOS A TIPO STR

In [276]:
# En esta celda lo que hacemos es llamar a la funcion que se encarga de realizar el cambio de tipo de dato a tipo str. Para eso en la siguiente linea
# lo que hacemos es crear una lista con las columnas que deseamos realizar dicho cambio.
columnas = ["order_id","customer_id","order_status"]
# Una vez que se tenga la lista con los campos deseado, en la siguiente linea llamamos a la funcion y guardamos en df_orders_mod, el dataframe con los
# cambios realizados por la funcion.
df_orders_mod = CambioTipoStr(df_orders_mod,columnas)

#### CAMBIO CAMPO A MINUSCULA

In [278]:
# En esta celda lo que se hace es llamar a la funcion que se encarga de hacer que todos los calores de un campo que contriene str, aparezcan
# con minuscula. Para eso en la siguiente linea lo que se hace es crear una lista con los campos que deseamos que aparezcan en minuscula.
columnas = ["order_status"]
# Una vez que ya tenemos la lista con los campos, lo que hacemos es llamar a la funcion que se encarga de eso, donde se le envia el dataframe
# y la lista con los campos que deseamos modificar.
df_orders_mod = CambioMinuscula(df_orders_mod,columnas)

In [283]:
df_orders_mod.dtypes
# con esta linea podemos visualizar cuales son los tipos de datos de los campos del dataframe que ya fue modificado,
# donde podemos ver que ahora tenemos los campos con los tipos de variables deseados

order_id                         string[python]
customer_id                      string[python]
order_status                     string[python]
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

In [237]:
df_orders_mod.isnull().sum()
# Con esta linea buscamos verificar que no existan mas valores nulos, y realmente es posible verificar que en eldataframe
# ya no existen valores nulos, que es lo que se buscaba

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [239]:
df_orders_mod.shape
# De esta manera podemos ver que el dataframe sigue conservando la cantidad de campos, pero ya no contiene la misma cantidad de registros,
# ya que se han eliminados varios registros.

(96461, 8)

In [241]:
df_orders_mod.head()
# en esta linea volvemos a visualizar los primeros 5 registros del dataframe modificado, con el fin de poder visualizar
# lo cambios realizados.

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

### DF_CUSTOMERS

#### CAMBIAMOS CAMPOS A TIPO STR

In [243]:
# En esta celda lo que buscamos hacer es cambiar el tipo de datos de algunas columnas en tipo str. En la siguiente linea
# creamos una lista con las columnas a las cuales neceitamos realizar dicho cambio
columnas = ["customer_id","customer_unique_id","customer_city","customer_state"]
# En la linea siguiente llamamos a la funcion encargada de dicha tarea, en donde enviamos el dataframe que deceamos modifical
# y tambien las columnas de dicho dataframe qye queremos cambiar
df_customers_mod = CambioTipoStr(df_customers,columnas)

#### CAMBIO CAMPO A MINUSCULA

In [245]:
# En la sguiente celda buscamos modificar columnas que son str y que aparezcan todos los registros escritos en minuscula
# en la linea siguiente creamos una lista con las columnas que necesitamos modificar
columnas = ["customer_city"]
# ahora llamamos a la funcion encargada de eso yle enviamos el dataframe que necesitamos modificar, con las
# columnas necesarias
df_customers_mod = CambioMinuscula(df_customers_mod,columnas)

#### CAMBIO CAMPO A MAYUSCULA

In [247]:
# En la sguiente celda buscamos modificar columnas que son str y que aparezcan todos los registros escritos en mayuscula
# en la linea siguiente creamos una lista con las columnas que necesitamos modificar
columnas = ["customer_state"]
# ahora llamamos a la funcion encargada de eso yle enviamos el dataframe que necesitamos modificar, con las
# columnas necesarias
df_customers_mod = CambioMayuscula(df_customers_mod,columnas)

In [249]:
# con dtypes vamos a controlar que se hayan realizado los cambios de tipo de datos. 
df_customers_mod.dtypes

customer_id                 string[python]
customer_unique_id          string[python]
customer_zip_code_prefix             int64
customer_city               string[python]
customer_state              string[python]
dtype: object

In [251]:
# visualmente controlamos los primeros 5 registros del dataframe, con el fin de podecon controlar si se hicieron
# los cambios anteriores
df_customers_mod.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

### DF_ORDER_ITEMS

#### CAMBIAMOS CAMPOS A TIPO DATATIME

In [253]:
# en esta celda llamamos a la funcion encargada de cambiar el tipo de dato de las columnas a tipo datatime
# en la siguiente linea creamos una lista con las columas que necesitamos realizar dicho cambio
columnas = ["shipping_limit_date"]
# ahora llamamos a la funcion y le enviamos el dataframe y las columnas que deseanis cambiar
df_order_items_mod = CambioTipoFecha(df_order_items,columnas)

#### CAMBIAMOS DATOS A TIPO STR

In [255]:
# en esta celda llamanos a la funcion encargada de cambiar el tipo de datos de las columnas a str
# en la siguiente line creamos una lista con las columnas que necesitamos cambiar
columnas = ["order_id","product_id","seller_id"]
# ahora llamamos a la funcion y le enviamos el dataframe necesario con las columnas que deseamos cambiar
df_order_items_mod = CambioTipoStr(df_order_items_mod,columnas)

In [257]:
# controlamos que los cambios ahora tengas los tipos de datos que necesitamos
df_order_items_mod.dtypes

order_id               string[python]
order_item_id                   int64
product_id             string[python]
seller_id              string[python]
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object

### DF_PRODUCTS

#### ANALIZAMOS VALORES NULOS

ESTE DATAFRAME POSEE 4 COLUMAS CON VALORES NULOS, LAS CUALES SON LAS SIGUINTES:

product_category_name         610

product_name_lenght           610

product_description_lenght    610

product_photos_qty            610

product_weight_g                2

product_length_cm               2

product_height_cm               2

product_width_cm                2

LA REGLA QUE SERA UTILIZADA EN EL MANEJO DE LOS NULOS ES LA SIGUIENTE:

PARA EL CAMPO "product_category_name" REEMPLAZAREMOS LOS CAMPOS NULS POR "Desconocida"

PARA EL RESTO DE LOS CAMPOS, REEMPLAZAREMOS LOS NULOS POR EL VALOR "-1", EL CUAL DICHO VALOR, INDICA QUE ES DESCONOCIDO EL VALOR REAL.

SE TOMO ESTA DECICION, YA QUE LOS CAMPOS CON LOS QUE ESTAMOS TRABAJANDO, NO SON DE VALORES DE PRECIO, NI CANTIDAD DE STOCK, NI OTROS VALORES 

DE ESE TIPO. LOS VALORES CON LOS QUE TRABAJAMOS EN ESTE DATAFRAME, SON TAMAÑO DE PRODUCTO, LARGO DEL NOMBRE, CANTIDAD DE IMAGENES EN EL POSTEO.

SE TENDRA EN CUENTA QUE AL MOMENTO DE QUERER REALIZAR ALGUN CALCULO, DE SER NECESARIO EXCLUIREMOS A LOS REGISTROS SON -1, YA QUE SON DESCONOCIDOS

In [259]:
# en la siguiente line vamos a completar los vlores nulos con la palabra "Desconocido"
df_products["product_category_name"] = df_products["product_category_name"].fillna("Desconocida")
# A continuacion creamos una lista con las columnas que vamos a querer modificar, luego a todas esas columnas la completamos
#en sus valores nulos con el valor -1
columnas = ["product_name_lenght","product_description_lenght","product_photos_qty","product_weight_g","product_length_cm","product_height_cm","product_width_cm"]
df_products[columnas] = df_products[columnas].fillna(-1)

#### CAMBIAMOS CAMPOS A TIPO STR

In [261]:
# en esta celda lo que hacemos es llamar a la funcion encargada de cambiar el tipo de dato de los campos a tipo str.
# en la siguiente linea creamos una lista con los campos que desemaos cambiar el tipo de dato a tipo str
columnas = ["product_id","product_category_name"]
# luego llamamos a la funcion que se encarda de dicha tarea y le eviamos el dataframe y las columnas que deseamos cambiar
df_products_mod = CambioTipoStr(df_products,columnas)

#### CAMBIAMOS CAMPO A MINUSCULA

In [263]:
# en esta celda estamos buscando que en los campos deseados del dataframe, nos aseguremos que aparezcan en minuscula.
# en la proxima linea creamos una lista con los campos deaseados a cambiar a minuscula
columnas = ["product_category_name"]
# ahora llamanos a la funcion encargada de dicha tarea y le enviamos el dataframe y las columnas que deseamos cambiar
df_products_mod = CambioMinuscula(df_products_mod,columnas)

In [265]:
# visualizamos que los campos ahora tengan los tipos de datos que necesitamos
df_products_mod.dtypes

product_id                    string[python]
product_category_name         string[python]
product_name_lenght                  float64
product_description_lenght           float64
product_photos_qty                   float64
product_weight_g                     float64
product_length_cm                    float64
product_height_cm                    float64
product_width_cm                     float64
dtype: object

In [267]:
#controlamos que la sumatoria de valores nulos en cada campo sea cero
df_products_mod.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [269]:
# vemos la estructura del dataframe y se observa que la cantidad de columnas sigue siendo la mismoa, al igual que los registros,
# ya que no se eliminaron registros, si no que se completaron con otros valres
df_products_mod.shape

(32951, 9)

In [271]:
# controlamos visualmente que los datos aparezcan como los hemos trabajado
df_products_mod.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

### DF_ORDER_PAYMENTS

#### CAMBIAMOS CAMPOS A TIPO STR

In [51]:
# en esta caleda llamamos a la funcion encargada de transformar el tipo de dato de los campos de un dataframe a tipo str
# en la siguiente linea creamos una lista con los campos que neceitamos mnodificar
columnas = ["order_id","payment_type"]
# ahora llamamos a la funcion, enviandole el dataframe y los campos necesarios
df_order_payments_mod = CambioTipoStr(df_order_payments,columnas)

#### CAMBIAMOS CAMPOS A MINUSCULA

In [ ]:
# en esta celda llamamos a la funcion encargada de hacer que losvalores de un campo aparezcan en minuscula
# en la linea que sigue creamos una lista con los campos que necesitamos cambiar
columnas = ["payment_type"]
# en la linea que sigue llamamos a la funcion y le enviamos el dataframe y los campos deseados
df_order_payments_mod = CambioMinuscula(df_order_payments_mod,columnas)

In [53]:
# ahora visualizamos que los tipos de datos de los campos esten como necesitamos.
df_order_payments_mod.dtypes

order_id                string[python]
payment_sequential               int64
payment_type            string[python]
payment_installments             int64
payment_value                  float64
dtype: object